### ISSUE TITLE
No filename information for errors in ES modules

### ISSUE THREAD

**Issue by fatcerberus** [Microsoft/ChakraCore#3618]
A runtime error in an ES module produces a stack trace like:

```
ReferenceError: 'getEaten' is not defined
   at Showcase.prototype.on_update (:49:3)
   at Anonymous function (#/game_modules/thread.js:134:26)
   at Anonymous function (#/game_modules/thread.js:195:3)
   at next (#/runtime/from.js:267:4)
   at nullOp (#/runtime/from.js:706:9)
   at Anonymous function (#/runtime/from.js:101:3)
   at _updateAll (#/game_modules/thread.js:187:2)
```

The frames above that have filename information were in scripts loaded using `require()`.  ES module frames like the top one `Showcase.prototype.on_update` do not.

**Comment by fatcerberus** [comment]
This also affects the debugging API - i.e. `JsDiagGetScripts()` returns an empty string for the filename of ES modules.

**Comment by digitalinfinity** [comment]
@fatcerberus thanks for reporting!

I can reproduce this in `ch` but I'm curious, how are you using ES modules in your scenario? The problem seems to be that JsParseModuleSource doesn't take in a `url` since it's meant to have arbitrary strings parsed in as a URL. Interestingly `JsExperimentalApiRunModule` does take a `sourceUrl` parameter that should get reflected correctly but I think that API is deprecated. I think the lack of a url parameter is probably on oversight. @liminzhu do you know? And do you know if we can just change this API to take in a url parameter- the API appears to be core only so I think this is an experimental API that's ok to break, but I'm not certain.

cc @Yongqu 

@fatcerberus once we get more clarity on what to do, the change itself should be relatively straightforward to make- if you want to take a stab at submitting a PR, you can message me on gitter and I'll be happy to walk you through what changes would be needed. If you prefer we make the change, that's perfectly fine too 😄

**Comment by fatcerberus** [comment]
Here's my module loader:
https://github.com/fatcerberus/minisphere/blob/ec269e213decde78a5d9a548336d643e0fddbd28/src/shared/jsal.c#L396-L444

And my fetch callback:
https://github.com/fatcerberus/minisphere/blob/ec269e213decde78a5d9a548336d643e0fddbd28/src/shared/jsal.c#L1818-L1868

**Comment by fatcerberus** [comment]
> how are you using ES modules in your scenario?

My application is a JavaScript game engine, miniSphere:
https://github.com/fatcerberus/minisphere

My engine was based on Duktape, I'm in the process of migrating to ChakraCore.  ES modules are the preferred medium for writing Sphere code, although I fully support `require()` and traditional JS as well.

Previously I was using Babel in order to support ESM but I'd like to take advantage of Chakra's in-built support and avoid the build step entirely.  The thing is that my tool chain also has a full debugger (SSj) so not being able to see filenames of modules through the debug protocol is a showstopper :(

**Comment by liminzhu** [comment]
@digitalinfinity I didn't design the module API, but if I have to guess sourceUrl may be unnecessary in `JsParseModuleSource`. We take in a specifier when initializing the module record with `JsInitializeModuleRecord`, so that info should be somewhere along the module record. @suwc do you know?

**Comment by fatcerberus** [comment]
I would like to fix this, but the Chakra codebase is multilayered and huge, I wouldn't even know where to start.

**Comment by fatcerberus** [comment]
This is a showstopper for me because the filenames are not visible through the debugging API either and my debugger then doesn't know which source file to open.  I've somewhat hacked around it by incrementing an internal script ID whenever I compile a script and using *that* to track the filename, but if a script calls `eval` or `new Function` then I get out of sync.

**Comment by fatcerberus** [comment]
It looks like the issue is:
https://github.com/Microsoft/ChakraCore/blob/09ab25c9a4e28dba22d2130ed323ab791676b11d/lib/Jsrt/Core/JsrtCore.cpp#L80

ParseModuleSource calls `CreateSourceContextInfo()` with the url set to `nullptr`.  This should actually be a pretty easy fix, I'll get a PR up soon :)

**Comment by rhuanjl** [comment]
I've been discussing this with @fatcerberus and exploring how the underlying code works.

The specifier is in the module record, but it's not passed to the CreateSourceContextInfo function, if you do pass it you fail tests - because with tracing enabled any URL given to CreateSourceContextInfo that doesn't actually lead to a file ends up here: https://github.com/Microsoft/ChakraCore/blob/master/lib/Runtime/Language/DynamicProfileStorage.h#L152-L160

One option is to set the specifier as the URL after CreateSourceContextInfo is done, feels like a dirty hack but passes the tests when I run it offline, basically doing this on lines 78-89 of JsrtCore.cpp
```cc


@liminzhu what would be your thought of adding an extra

**Comment by rhuanjl** [comment]
I've been discussing this with @fatcerberus and exploring how the underlying code works.

The specifier is in the module record, but it's not passed to the CreateSourceContextInfo function, if you do pass it you fail tests - because with tracing enabled (i.e. in debug/test mode) any URL given to CreateSourceContextInfo that doesn't actually lead to a file ends up here: https://github.com/Microsoft/ChakraCore/blob/master/lib/Runtime/Language/DynamicProfileStorage.h#L152-L160

**One option is to set the specifier as the URL after CreateSourceContextInfo is done, this feels like a dirty hack but passes the tests when I run it offline, basically doing this on lines 78-89 of JsrtCore.cpp (inJsParseModuleSource)**
```cc
if (sourceContextInfo == nullptr)
{
	//create SourceContextInfo with null URL and set specifier as URL afterwards
	sourceContextInfo = scriptContext->CreateSourceContextInfo(sourceContext, nullptr, 0, nullptr, nullptr, 0);
	Js::JavascriptString *specifier = Js::JavascriptString::FromVar(moduleRecord->GetSpecifier());
	const char16 *urlSz = specifier->GetSz();
	size_t urlLength = specifier->GetLength() + 1;
			
	char16* moduleURLName = AnewArray(scriptContext->SourceCodeAllocator(), char16, urlLength);
	js_wmemcpy_s(moduleURLName, urlLength, urlSz, urlLength -1);
	moduleURLName[urlLength - 1] = _u('\0');
	sourceContextInfo->url = moduleURLName;
}
```
**Problems with this are:**
a) it doesn't enable LoadFromDynamicProfileStorage - I don't know what this does but assume it should be a medium/long term goal to enable it? https://github.com/Microsoft/ChakraCore/blob/master/lib/Runtime/Base/ScriptContext.cpp#L2537-L2545

b) It's still the specifier not a URL - some hosts may supply the URL as the specifier but they may not - and so you're left with a SourceContextInfo record with a property that isn't necessarily what it's name implies - this could result in future errors/problems.

**Any other option looks like it would involve changing a parameter for an API or adding a new API**
A simple option if you didn't mind adding a new API would be an optional function that could be called before JsParseModuleSource to set a URL onto the module record - if it was called that URL would then be passed on during JsParseModuleSource. Any other option looks like it would involve changing a parameter as as far as I can see there is no module function that takes an actual URL at the moment.
@liminzhu what would be your thought of adding an extra API for this?

**Comment by fatcerberus** [comment]
This was actually already fixed by #4080, so I'll close this 😃
